File: src/server/content/adminHttp.test.ts
import { describe, expect, it, vi } from 'vitest'

import {
  ContentAdminNotFoundError,
  ContentAdminValidationError,
  NoUnpublishedChangesError,
  PublishedRevisionNotFoundError,
} from './adminRepository'
import { createAdminHttpHandlers, createLazyDatabaseAdminHttpHandlers } from './adminHttp'

const adminAuthEnv = {
  CONTENT_ADMIN_USERNAME: 'editor',
  CONTENT_ADMIN_PASSWORD: 'secret',
}

function createResponseRecorder() {
  return {
    statusCode: 200,
    headers: {} as Record<string, string>,
    body: undefined as unknown,
    status(code: number) {
      this.statusCode = code
      return this
    },
    setHeader(name: string, value: string) {
      this.headers[name] = value
    },
    json(value: unknown) {
      this.body = value
      return this
    },
  }
}

describe('content admin HTTP handlers', () => {
  it('lists lessons or returns one lesson snapshot from the shared lessons endpoint', async () => {
    const snapshot = {
      lessonId: 'self-intro',
      slug: 'self-intro',
      displayOrder: 1,
      enabled: true,
      draftLesson: { id: 'self-intro' },
      publishedLesson: { id: 'self-intro' },
      modules: [],
      publishedHistory: {
        lessonMeta: [],
        dialogue: [],
        sentencePatterns: [],
        vocabulary: [],
        pronunciation: [],
        hanziRecognition: [],
        practice: [],
        reviewCards: [],
        shortInput: [],
      },
    }
    const repository = {
      listLessons: vi.fn().mockResolvedValue([{ lessonId: 'self-intro', slug: 'self-intro', displayOrder: 1, enabled: true, draftChangedModuleCount: 1 }]),
      getLessonSnapshot: vi.fn().mockResolvedValue(snapshot),
      saveDraftModule: vi.fn(),
      publishModule: vi.fn(),
      rollbackModule: vi.fn(),
    }
    const handlers = createAdminHttpHandlers(repository, adminAuthEnv)

    const listResponse = createResponseRecorder()
    await handlers.lessons({ method: 'GET', query: {}, headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' } }, listResponse)
    expect(listResponse.statusCode).toBe(200)
    expect(repository.listLessons).toHaveBeenCalledTimes(1)

    const detailResponse = createResponseRecorder()
    await handlers.lessons(
      {
        method: 'GET',
        query: { lessonId: 'self-intro' },
        headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' },
      },
      detailResponse,
    )
    expect(detailResponse.statusCode).toBe(200)
    expect(detailResponse.body).toEqual(snapshot)
    expect(repository.getLessonSnapshot).toHaveBeenCalledWith('self-intro')
  })

  it('saves drafts, publishes modules, and rolls back published revisions from JSON request bodies', async () => {
    const repository = {
      listLessons: vi.fn(),
      getLessonSnapshot: vi.fn(),
      saveDraftModule: vi.fn().mockResolvedValue({ ok: 'draft' }),
      publishModule: vi.fn().mockResolvedValue({ ok: 'publish' }),
      rollbackModule: vi.fn().mockResolvedValue({ ok: 'rollback' }),
    }
    const handlers = createAdminHttpHandlers(repository, adminAuthEnv)

    const draftResponse = createResponseRecorder()
    await handlers.draft(
      {
        method: 'PUT',
        headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' },
        body: JSON.stringify({
          lessonId: 'self-intro',
          moduleType: 'lessonMeta',
          payload: { id: 'self-intro', title: { en: 'Draft', fr: 'Brouillon' }, scenario: { en: 'Scenario', fr: 'Scénario' } },
          note: 'Save title draft',
        }),
      },
      draftResponse,
    )
    expect(draftResponse.statusCode).toBe(200)
    expect(repository.saveDraftModule).toHaveBeenCalledWith(
      expect.objectContaining({ lessonId: 'self-intro', moduleType: 'lessonMeta', note: 'Save title draft' }),
    )

    const publishResponse = createResponseRecorder()
    await handlers.publish(
      {
        method: 'POST',
        headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' },
        body: { lessonId: 'self-intro', moduleType: 'lessonMeta', note: 'Publish title draft' },
      },
      publishResponse,
    )
    expect(publishResponse.statusCode).toBe(200)
    expect(repository.publishModule).toHaveBeenCalledWith(
      expect.objectContaining({ lessonId: 'self-intro', moduleType: 'lessonMeta', note: 'Publish title draft' }),
    )

    const rollbackResponse = createResponseRecorder()
    await handlers.rollback(
      {
        method: 'POST',
        headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' },
        body: { lessonId: 'self-intro', moduleType: 'lessonMeta', publishedRevisionId: 901, note: 'Rollback title draft' },
      },
      rollbackResponse,
    )
    expect(rollbackResponse.statusCode).toBe(200)
    expect(repository.rollbackModule).toHaveBeenCalledWith(
      expect.objectContaining({ lessonId: 'self-intro', moduleType: 'lessonMeta', publishedRevisionId: 901 }),
    )
  })

  it('maps admin errors to safe HTTP statuses and keeps method guards strict', async () => {
    const repository = {
      listLessons: vi.fn().mockRejectedValue(new ContentAdminNotFoundError('missing lesson')),
      getLessonSnapshot: vi.fn().mockRejectedValue(new ContentAdminNotFoundError('missing lesson')),
      saveDraftModule: vi.fn().mockRejectedValue(new ContentAdminValidationError('Invalid lessonMeta payload')),
      publishModule: vi.fn().mockRejectedValue(new NoUnpublishedChangesError('self-intro', 'lessonMeta')),
      rollbackModule: vi.fn().mockRejectedValue(new PublishedRevisionNotFoundError('self-intro', 'lessonMeta', 404)),
    }
    const handlers = createAdminHttpHandlers(repository, adminAuthEnv)

    const methodNotAllowed = createResponseRecorder()
    await handlers.lessons({ method: 'POST', query: {}, headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' } }, methodNotAllowed)
    expect(methodNotAllowed.statusCode).toBe(405)
    expect(methodNotAllowed.headers.Allow).toBe('GET')

    const badDraft = createResponseRecorder()
    await handlers.draft({ method: 'PUT', headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' }, body: {} }, badDraft)
    expect(badDraft.statusCode).toBe(400)

    const missingLesson = createResponseRecorder()
    await handlers.lessons(
      {
        method: 'GET',
        query: { lessonId: 'missing' },
        headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' },
      },
      missingLesson,
    )
    expect(missingLesson.statusCode).toBe(404)

    const invalidDraft = createResponseRecorder()
    await handlers.draft(
      {
        method: 'PUT',
        headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' },
        body: { lessonId: 'self-intro', moduleType: 'lessonMeta', payload: { id: 'self-intro' } },
      },
      invalidDraft,
    )
    expect(invalidDraft.statusCode).toBe(400)

    const noChanges = createResponseRecorder()
    await handlers.publish(
      {
        method: 'POST',
        headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' },
        body: { lessonId: 'self-intro', moduleType: 'lessonMeta' },
      },
      noChanges,
    )
    expect(noChanges.statusCode).toBe(409)

    const missingRevision = createResponseRecorder()
    await handlers.rollback(
      {
        method: 'POST',
        headers: { authorization: 'Basic ZWRpdG9yOnNlY3JldA==' },
        body: { lessonId: 'self-intro', moduleType: 'lessonMeta', publishedRevisionId: 404 },
      },
      missingRevision,
    )
    expect(missingRevision.statusCode).toBe(404)
  })

  it('returns 503 from the lazy database handlers when MySQL env is unavailable', async () => {
    const handlers = createLazyDatabaseAdminHttpHandlers({})
    const response = createResponseRecorder()

    await handlers.lessons({ method: 'GET', query: {} }, response)

    expect(response.statusCode).toBe(503)
    expect(response.body).toEqual({ error: 'Content admin database is not configured' })
  })

  it('rejects unauthenticated admin requests with a 401 challenge before reaching the repository', async () => {
    const repository = {
      listLessons: vi.fn(),
      getLessonSnapshot: vi.fn(),
      saveDraftModule: vi.fn(),
      publishModule: vi.fn(),
      rollbackModule: vi.fn(),
    }
    const handlers = createAdminHttpHandlers(repository, adminAuthEnv)
    const response = createResponseRecorder()

    await handlers.lessons({ method: 'GET', query: {}, headers: {} }, response)

    expect(response.statusCode).toBe(401)
    expect(response.headers['WWW-Authenticate']).toContain('Basic')
    expect(repository.listLessons).not.toHaveBeenCalled()
  })

  it('returns a plain 401 without a browser auth challenge for admin UI fetch requests', async () => {
    const repository = {
      listLessons: vi.fn(),
      getLessonSnapshot: vi.fn(),
      saveDraftModule: vi.fn(),
      publishModule: vi.fn(),
      rollbackModule: vi.fn(),
    }
    const handlers = createAdminHttpHandlers(repository, adminAuthEnv)
    const response = createResponseRecorder()

    await handlers.lessons(
      {
        method: 'GET',
        query: {},
        headers: { 'x-content-admin-client': 'spa' },
      },
      response,
    )

    expect(response.statusCode).toBe(401)
    expect(response.headers['WWW-Authenticate']).toBeUndefined()
    expect(response.body).toEqual({ error: 'Admin authentication required' })
    expect(repository.listLessons).not.toHaveBeenCalled()
  })

  it('accepts authenticated admin requests and returns 400 for malformed JSON bodies instead of 500', async () => {
    const repository = {
      listLessons: vi.fn().mockResolvedValue([{ lessonId: 'self-intro' }]),
      getLessonSnapshot: vi.fn(),
      saveDraftModule: vi.fn(),
      publishModule: vi.fn(),
      rollbackModule: vi.fn(),
    }
    const handlers = createAdminHttpHandlers(repository, adminAuthEnv)
    const authHeaders = {
      authorization: 'Basic ZWRpdG9yOnNlY3JldA==',
    }

    const okResponse = createResponseRecorder()
    await handlers.lessons({ method: 'GET', query: {}, headers: authHeaders }, okResponse)
    expect(okResponse.statusCode).toBe(200)
    expect(repository.listLessons).toHaveBeenCalledTimes(1)

    const badJsonResponse = createResponseRecorder()
    await handlers.draft(
      {
        method: 'PUT',
        headers: authHeaders,
        body: '{"lessonId":"self-intro"',
      },
      badJsonResponse,
    )
    expect(badJsonResponse.statusCode).toBe(400)
    expect(badJsonResponse.body).toEqual({ error: 'Invalid JSON request body' })
  })
})

File: internal/adminhttp/adminhttp_test.go
package adminhttp

import (
	"context"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"en-fr-chinese-entry-mvp/internal/adminrepo"
	"en-fr-chinese-entry-mvp/internal/auth"
)

type fakeRepo struct {
	listLessons       func(ctx context.Context) ([]adminrepo.AdminLessonSummary, error)
	getLessonSnapshot func(ctx context.Context, lessonID string) (*adminrepo.AdminLessonSnapshot, error)
	saveDraftModule   func(ctx context.Context, input adminrepo.SaveDraftModuleInput) (*adminrepo.AdminLessonSnapshot, error)
	publishModule     func(ctx context.Context, input adminrepo.PublishModuleInput) (*adminrepo.AdminLessonSnapshot, error)
	rollbackModule    func(ctx context.Context, input adminrepo.RollbackModuleInput) (*adminrepo.AdminLessonSnapshot, error)
	listLessonsCalls  int
}

func (f *fakeRepo) ListLessons(ctx context.Context) ([]adminrepo.AdminLessonSummary, error) {
	f.listLessonsCalls++
	if f.listLessons != nil {
		return f.listLessons(ctx)
	}
	return []adminrepo.AdminLessonSummary{{LessonID: "self-intro", Slug: "self-intro", DisplayOrder: 1, Enabled: true, DraftChangedModuleCount: 1}}, nil
}

func (f *fakeRepo) GetLessonSnapshot(ctx context.Context, lessonID string) (*adminrepo.AdminLessonSnapshot, error) {
	if f.getLessonSnapshot != nil {
		return f.getLessonSnapshot(ctx, lessonID)
	}
	return &adminrepo.AdminLessonSnapshot{LessonID: lessonID}, nil
}

func (f *fakeRepo) SaveDraftModule(ctx context.Context, input adminrepo.SaveDraftModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
	if f.saveDraftModule != nil {
		return f.saveDraftModule(ctx, input)
	}
	return &adminrepo.AdminLessonSnapshot{LessonID: input.LessonID}, nil
}

func (f *fakeRepo) PublishModule(ctx context.Context, input adminrepo.PublishModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
	if f.publishModule != nil {
		return f.publishModule(ctx, input)
	}
	return &adminrepo.AdminLessonSnapshot{LessonID: input.LessonID}, nil
}

func (f *fakeRepo) RollbackModule(ctx context.Context, input adminrepo.RollbackModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
	if f.rollbackModule != nil {
		return f.rollbackModule(ctx, input)
	}
	return &adminrepo.AdminLessonSnapshot{LessonID: input.LessonID}, nil
}

var adminAuthEnv = auth.Env{ContentAdminUsername: "editor", ContentAdminPassword: "secret"}

const basicAuth = "Basic ZWRpdG9yOnNlY3JldA=="

func newTestHandler(repo Repository, env auth.Env) http.Handler {
	return New(repo, env)
}

func doRequest(t *testing.T, handler http.Handler, method, path string, headers map[string]string, body string) *httptest.ResponseRecorder {
	t.Helper()
	var reader io.Reader
	if body != "" {
		reader = strings.NewReader(body)
	}
	req := httptest.NewRequest(method, path, reader)
	for key, value := range headers {
		req.Header.Set(key, value)
	}
	rec := httptest.NewRecorder()
	handler.ServeHTTP(rec, req)
	return rec
}

func TestLessonsListAndDetail(t *testing.T) {
	repo := &fakeRepo{}
	handler := newTestHandler(repo, adminAuthEnv)

	list := doRequest(t, handler, http.MethodGet, "/api/admin/content/lessons", map[string]string{"Authorization": basicAuth}, "")
	if list.Code != http.StatusOK {
		t.Fatalf("list status = %d, want 200", list.Code)
	}
	if repo.listLessonsCalls != 1 {
		t.Fatalf("listLessons calls = %d, want 1", repo.listLessonsCalls)
	}

	detail := doRequest(t, handler, http.MethodGet, "/api/admin/content/lessons?lessonId=self-intro", map[string]string{"Authorization": basicAuth}, "")
	if detail.Code != http.StatusOK {
		t.Fatalf("detail status = %d, want 200", detail.Code)
	}
	if !strings.Contains(detail.Body.String(), `"lessonId":"self-intro"`) {
		t.Fatalf("detail body = %s", detail.Body.String())
	}
}

func TestLessonsRetriesTransientTimeoutOnce(t *testing.T) {
	transient := &adminConnError{message: "connect ETIMEDOUT"}
	calls := 0
	repo := &fakeRepo{
		listLessons: func(ctx context.Context) ([]adminrepo.AdminLessonSummary, error) {
			calls++
			if calls == 1 {
				return nil, transient
			}
			return []adminrepo.AdminLessonSummary{{LessonID: "self-intro"}}, nil
		},
	}
	handler := newTestHandler(repo, adminAuthEnv)

	rec := doRequest(t, handler, http.MethodGet, "/api/admin/content/lessons", map[string]string{"Authorization": basicAuth}, "")
	if rec.Code != http.StatusOK {
		t.Fatalf("status = %d, want 200", rec.Code)
	}
	if calls != 2 {
		t.Fatalf("listLessons calls = %d, want 2 (one retry)", calls)
	}
}

func TestDraftPublishRollback(t *testing.T) {
	var saved *adminrepo.SaveDraftModuleInput
	var published *adminrepo.PublishModuleInput
	var rolledBack *adminrepo.RollbackModuleInput
	repo := &fakeRepo{
		saveDraftModule: func(ctx context.Context, input adminrepo.SaveDraftModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
			saved = &input
			return &adminrepo.AdminLessonSnapshot{LessonID: input.LessonID}, nil
		},
		publishModule: func(ctx context.Context, input adminrepo.PublishModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
			published = &input
			return &adminrepo.AdminLessonSnapshot{LessonID: input.LessonID}, nil
		},
		rollbackModule: func(ctx context.Context, input adminrepo.RollbackModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
			rolledBack = &input
			return &adminrepo.AdminLessonSnapshot{LessonID: input.LessonID}, nil
		},
	}
	handler := newTestHandler(repo, adminAuthEnv)

	draftBody := `{"lessonId":"self-intro","moduleType":"lessonMeta","payload":{"id":"self-intro"},"note":"Save title draft"}`
	draft := doRequest(t, handler, http.MethodPut, "/api/admin/content/draft", map[string]string{"Authorization": basicAuth}, draftBody)
	if draft.Code != http.StatusOK {
		t.Fatalf("draft status = %d, want 200", draft.Code)
	}
	if saved == nil || saved.LessonID != "self-intro" || saved.ModuleType != "lessonMeta" || saved.Note == nil || *saved.Note != "Save title draft" {
		t.Fatalf("saved = %+v", saved)
	}
	if saved.CreatedBy != "admin-ui" {
		t.Fatalf("createdBy = %q, want admin-ui", saved.CreatedBy)
	}

	publishBody := `{"lessonId":"self-intro","moduleType":"lessonMeta","note":"Publish title draft"}`
	pub := doRequest(t, handler, http.MethodPost, "/api/admin/content/publish", map[string]string{"Authorization": basicAuth}, publishBody)
	if pub.Code != http.StatusOK {
		t.Fatalf("publish status = %d, want 200", pub.Code)
	}
	if published == nil || published.LessonID != "self-intro" || published.Note == nil || *published.Note != "Publish title draft" {
		t.Fatalf("published = %+v", published)
	}

	rollbackBody := `{"lessonId":"self-intro","moduleType":"lessonMeta","publishedRevisionId":901,"note":"Rollback title draft"}`
	rb := doRequest(t, handler, http.MethodPost, "/api/admin/content/rollback", map[string]string{"Authorization": basicAuth}, rollbackBody)
	if rb.Code != http.StatusOK {
		t.Fatalf("rollback status = %d, want 200", rb.Code)
	}
	if rolledBack == nil || rolledBack.PublishedRevisionID != 901 {
		t.Fatalf("rolledBack = %+v", rolledBack)
	}
}

func TestErrorMappingAndMethodGuards(t *testing.T) {
	repo := &fakeRepo{
		getLessonSnapshot: func(ctx context.Context, lessonID string) (*adminrepo.AdminLessonSnapshot, error) {
			return nil, &adminrepo.NotFoundError{Message: "Editable lesson not found: " + lessonID}
		},
		saveDraftModule: func(ctx context.Context, input adminrepo.SaveDraftModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
			return nil, &adminrepo.ValidationError{Message: "Invalid lessonMeta payload"}
		},
		publishModule: func(ctx context.Context, input adminrepo.PublishModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
			return nil, adminrepo.ErrNoUnpublishedChanges
		},
		rollbackModule: func(ctx context.Context, input adminrepo.RollbackModuleInput) (*adminrepo.AdminLessonSnapshot, error) {
			return nil, adminrepo.ErrPublishedRevisionNotFound
		},
	}
	handler := newTestHandler(repo, adminAuthEnv)

	methodNotAllowed := doRequest(t, handler, http.MethodPost, "/api/admin/content/lessons", map[string]string{"Authorization": basicAuth}, "")
	if methodNotAllowed.Code != http.StatusMethodNotAllowed {
		t.Fatalf("method not allowed = %d, want 405", methodNotAllowed.Code)
	}
	if allow := methodNotAllowed.Header().Get("Allow"); allow != "GET" {
		t.Fatalf("Allow = %q, want GET", allow)
	}

	badDraft := doRequest(t, handler, http.MethodPut, "/api/admin/content/draft", map[string]string{"Authorization": basicAuth}, `{}`)
	if badDraft.Code != http.StatusBadRequest {
		t.Fatalf("badDraft = %d, want 400", badDraft.Code)
	}

	missingLesson := doRequest(t, handler, http.MethodGet, "/api/admin/content/lessons?lessonId=missing", map[string]string{"Authorization": basicAuth}, "")
	if missingLesson.Code != http.StatusNotFound {
		t.Fatalf("missingLesson = %d, want 404", missingLesson.Code)
	}

	invalidDraft := doRequest(t, handler, http.MethodPut, "/api/admin/content/draft", map[string]string{"Authorization": basicAuth}, `{"lessonId":"self-intro","moduleType":"lessonMeta","payload":{"id":"self-intro"}}`)
	if invalidDraft.Code != http.StatusBadRequest {
		t.Fatalf("invalidDraft = %d, want 400", invalidDraft.Code)
	}

	noChanges := doRequest(t, handler, http.MethodPost, "/api/admin/content/publish", map[string]string{"Authorization": basicAuth}, `{"lessonId":"self-intro","moduleType":"lessonMeta"}`)
	if noChanges.Code != http.StatusConflict {
		t.Fatalf("noChanges = %d, want 409", noChanges.Code)
	}

	missingRevision := doRequest(t, handler, http.MethodPost, "/api/admin/content/rollback", map[string]string{"Authorization": basicAuth}, `{"lessonId":"self-intro","moduleType":"lessonMeta","publishedRevisionId":404}`)
	if missingRevision.Code != http.StatusNotFound {
		t.Fatalf("missingRevision = %d, want 404", missingRevision.Code)
	}
}

func TestAuthRequired(t *testing.T) {
	repo := &fakeRepo{}
	handler := newTestHandler(repo, adminAuthEnv)

	noAuth := doRequest(t, handler, http.MethodGet, "/api/admin/content/lessons", map[string]string{}, "")
	if noAuth.Code != http.StatusUnauthorized {
		t.Fatalf("noAuth = %d, want 401", noAuth.Code)
	}
	if !strings.Contains(noAuth.Header().Get("WWW-Authenticate"), "Basic") {
		t.Fatalf("WWW-Authenticate = %q, want Basic challenge", noAuth.Header().Get("WWW-Authenticate"))
	}
	if repo.listLessonsCalls != 0 {
		t.Fatal("repository should not be reached without auth")
	}

	spa := doRequest(t, handler, http.MethodGet, "/api/admin/content/lessons", map[string]string{"X-Content-Admin-Client": "spa"}, "")
	if spa.Code != http.StatusUnauthorized {
		t.Fatalf("spa = %d, want 401", spa.Code)
	}
	if spa.Header().Get("WWW-Authenticate") != "" {
		t.Fatalf("spa WWW-Authenticate = %q, want empty", spa.Header().Get("WWW-Authenticate"))
	}
}

func TestAuthNotConfiguredReturns503(t *testing.T) {
	repo := &fakeRepo{}
	handler := newTestHandler(repo, auth.Env{})

	rec := doRequest(t, handler, http.MethodGet, "/api/admin/content/lessons", map[string]string{"Authorization": basicAuth}, "")
	if rec.Code != http.StatusServiceUnavailable {
		t.Fatalf("status = %d, want 503", rec.Code)
	}
	if !strings.Contains(rec.Body.String(), "authentication is not configured") {
		t.Fatalf("body = %s", rec.Body.String())
	}
}

func TestMalformedJSONReturns400(t *testing.T) {
	repo := &fakeRepo{}
	handler := newTestHandler(repo, adminAuthEnv)

	rec := doRequest(t, handler, http.MethodPut, "/api/admin/content/draft", map[string]string{"Authorization": basicAuth}, `{"lessonId":"self-intro"`)
	if rec.Code != http.StatusBadRequest {
		t.Fatalf("status = %d, want 400", rec.Code)
	}
	if !strings.Contains(rec.Body.String(), "Invalid JSON request body") {
		t.Fatalf("body = %s", rec.Body.String())
	}
}

func TestUnexpectedErrorReturns500(t *testing.T) {
	repo := &fakeRepo{
		listLessons: func(ctx context.Context) ([]adminrepo.AdminLessonSummary, error) {
			return nil, errors.New("unexpected boom")
		},
	}
	handler := newTestHandler(repo, adminAuthEnv)

	rec := doRequest(t, handler, http.MethodGet, "/api/admin/content/lessons", map[string]string{"Authorization": basicAuth}, "")
	if rec.Code != http.StatusInternalServerError {
		t.Fatalf("status = %d, want 500", rec.Code)
	}
	if !strings.Contains(rec.Body.String(), "Unable to process content admin request") {
		t.Fatalf("body = %s", rec.Body.String())
	}
}

func TestUnknownPathReturns404(t *testing.T) {
	handler := newTestHandler(&fakeRepo{}, adminAuthEnv)
	rec := doRequest(t, handler, http.MethodGet, "/api/admin/content/nope", map[string]string{"Authorization": basicAuth}, "")
	if rec.Code != http.StatusNotFound {
		t.Fatalf("status = %d, want 404", rec.Code)
	}
}

func TestSanitizeDiagnosticText(t *testing.T) {
	input := "connect ETIMEDOUT mysql://admin:super-secret@db.example/content_admin?token=abc123 Authorization: Bearer xyz.abc-123"
	sanitized := sanitizeDiagnosticText(input)
	if strings.Contains(sanitized, "super-secret") || strings.Contains(sanitized, "abc123") || strings.Contains(sanitized, "xyz.abc-123") {
		t.Fatalf("sanitized leaked secret: %s", sanitized)
	}
	if !strings.Contains(sanitized, "[redacted]") {
		t.Fatalf("sanitized = %s, want [redacted]", sanitized)
	}
}
